src/main/java/org/apache/maven/plugin/compiler/TestCompilerMojo.java

```diff
@@ -369,9 +369,11 @@ else if ( Double.valueOf( getTarget() ) < Double.valueOf( MODULE_INFO_TARGET ) )
                 // very odd
                 // Means that main sources must be compiled with -modulesource and -Xmodule:<moduleName>
                 // However, this has a huge impact since you can't simply use it as a classpathEntry 
-                // due to extra folder in between
-                throw new UnsupportedOperationException( "Can't compile test sources "
-                    + "when main sources are missing a module descriptor" );
+                // due to extra folder in between ... or the user wants to test his main code on the
+                // classpath used by a named module ... or ...
+                getLog().warn( "Modular black-box test compilation with no main module?!" );
+
+                // TODO classpathElements = List.of( mainOutputDirectory );
```

Comment:
Todo? Can you link a JIRA? In order notbto forget?

Reply:
> "In order not to forget"

This. I'm not sure what to do here, actually. That's why I implemented the `throws` variant first. Now, a warning is emitted, and any subsequent "mis-configuration" will lead to an error. Presumably when calling `javac`... but, as stated in the comment, this "main=classic + test=module" case is rare and not well understood.

Reply:
@sormuras you know what I'm going to say ;) documentation is key here. Without having clear examples what's happening here and why somebody wants to do this I'd prefer to keep the Exception.

Reply:
I already excluded that case from an example project: https://github.com/micromata/sawdust/blob/master/pom.xml#L15

But, I might have found a valid use-case today: say you want to test an external module (on the module path) without writing main code on your own.